PurchasesTests/Networking/HTTPClientTests.swift

```diff
@@ -400,10 +400,117 @@ class HTTPClientTests: XCTestCase {
+        }
+
+        let totalRequests = Int.random(in: 50..<100)
+        for requestNumber in 0..<totalRequests {
+            self.client.performRequest("POST",
+                                       serially: true,
+                                       path: path,
+                                       body: ["requestNumber": requestNumber],
+                                       headers: nil) { (status, data, error) in
+                completionCallCount += 1
+            }
+        }
+        expect(completionCallCount).toEventually(equal(totalRequests))
+    }
+
+    func testPerformSerialRequestWaitsUntilFirstRequestIsDoneBeforeStartingSecond() {
+        let path = "/a_random_path"
+        var firstRequestFinished = false
+        var secondRequestFinished = false
+
+        stub(condition: isPath("/v1" + path)) { request in
+            usleep(30)
+            let requestData = request.ohhttpStubs_httpBody!
+            let requestBodyDict = try! JSONSerialization.jsonObject(with: requestData, options: []) as! [String: Any]
+
+            let requestNumber = requestBodyDict["requestNumber"] as! Int
+            if requestNumber == 2 {
+                expect(firstRequestFinished) == true
+            }
+
+            let json = "{\"message\": \"something is great up in the cloud\"}"
+            return HTTPStubsResponse(data: json.data(using: String.Encoding.utf8)!, statusCode:200, headers:nil)
+                .requestTime(0.1, responseTime: 0)
+        }
+
+        self.client.performRequest("POST",
+                                   serially: true,
+                                   path: path,
+                                   body: ["requestNumber": 1],
+                                   headers: nil) { (status, data, error) in
+            firstRequestFinished = true
+        }
+
+        self.client.performRequest("POST",
+                                   serially: true,
+                                   path: path,
+                                   body: ["requestNumber": 2],
+                                   headers: nil) { (status, data, error) in
+            secondRequestFinished = true
+        }
+
+        expect(firstRequestFinished).toEventually(beTrue())
+        expect(secondRequestFinished).toEventually(beTrue())
+    }
+
+    func testPerformConcurrentRequestDoesntWaitUntilFirstRequestIsDoneBeforeStartingSecond() {
+        let path = "/a_random_path"
+        var firstRequestFinished = false
+        var secondRequestFinished = false
+
+        stub(condition: isPath("/v1" + path)) { request in
+            let requestData = request.ohhttpStubs_httpBody!
+            let requestBodyDict = try! JSONSerialization.jsonObject(with: requestData, options: []) as! [String: Any]
+
+            let requestNumber = requestBodyDict["requestNumber"] as! Int
+            if requestNumber == 2 {
+                expect(firstRequestFinished) == false
```

Comment:
If both requests are concurrent, can you guarantee this is true? It's possible the first request has actually finished right?

Reply:
the way this test works is: 
- the code in `stub` runs when a request starts
- the code in the `performRequest` handler runs when a request ends
- in `stub`, I'm manually adding a 0.1 sec delay in the server response to give enough time for the other request to get started. 

I'm actually pretty sure that the delay is not even needed - the response will be async, and the code is going to start request 1, start request 2 on the same thread, then process whichever request arrived first once the thread is free